### DISCUSSION TITLE
Why 10 bytes for slots? Why 15 bits for payload_len?

### DISCUSSION THREAD

**Discussion by imoldfella** [leanstore/leanstore#3] [General]
Could someone explain what's going on here? Compiler Explorer declares it to be 10 bytes, seems like the 32 bit head is not going to be aligned? I'm curious why this structure vs say 6 bytes for the prefix or even pushing the payload_len into the data area? Thanks! I keep looking at and wonder what I'm missing. 

 struct __attribute__((packed)) Slot {
      // Layout:  key wihtout prefix | Payload
      u16 offset;
      u16 key_len;
      u16 payload_len : 15;
      union {
         HeadType head;
         u8 head_bytes[4];
      };
   };

**Comment by imoldfella** [comment]
Also, curious about the 4 byte key prefix being in the slot and in the data area. Simplification or optimization?

**Comment by adnan-alhomssi** [comment]
The 4 bytes key head is indeed stored twice: once in the slot to accelerate search and once in the data area to simplify the code mainly.
The reason for 15 bits payload_len is not there anymore and I just forgot to change it back, thanks for the catch!

**Comment by viktorleis** [comment]
You are also right that head is not aligned. On x86 unaligned loads are fast, so it's not that bad. We actually had an implementation that stored the length in the payload as you suggest, but it made the code more a bit more complicated.
